aip/0165.md

```diff
@@ -0,0 +1,102 @@
+pattern ([AIP-235][]) becomes infeasible.
+
+## Guidance
+
+**Important:** Most APIs **should** use only Delete ([AIP-135][]) or Batch
+Delete ([AIP-235][]) for deleting resources, and **should not** implement
+deleting based on criteria. This is because deleting is generally irreversible
+and it is possible for a user to irrecovably lose data with a careless query.
+
+An API **may** implement a Purge method to permit deleting a large number of
+resources based on a filter string; however, this **should** only be done if
+the Batch Delete ([AIP-235][]) pattern is insufficient to achieve the use case:
+
+```proto
+rpc PurgeBooks(PurgeBooksRequest) returns (PurgeBooksResponse) {
+  option (google.api.http) = {
+    post: "/v1/{parent=publishers/*}/books:purge"
+    body: "*"
+  }
+}
+```
+
+- The HTTP verb **must** be `POST`, and the `body` **must** be `"*"`.
+- The `parent` field **should** be included in the URI. If the API wishes to
+  support deletion across multiple parents, it **should** accept the `-`
+  character consistent with [AIP-159][].
+
+### Request message
+
+Purge methods implement a common request message pattern:
+
+```proto
+message PurgeBooksRequest {
+  // The publisher to purge books from.
+  // To purge books across publishers, send "publishers/-".
+  string parent = 1 [
+    (google.api.field_behavior) = REQUIRED,
+    (google.api.resource_reference) = {
+      child_type: "library-example.googleapis.com/Book"
+    }];
+
+  // A filter matching the books to be purged.
+  string filter = 2 [(google.api.field_behavior) = REQUIRED];
+
+  // Actually perform the purge.
+  // If `force` is set to false, the method will return a sample of
+  // resource names that would be deleted.
+  bool force = 3;
+}
+```
+
+- The `filter` field **should** be required, and **must** follow the same
+  semantics as in List methods (see [AIP-132][]).
+- The `force` field **must** be included. If it is not set, the API **must**
+  return a count of the resources that would be deleted as well as a sample of
+  those resources, without actually performing the deletion.
+
+### Response message
+
+Purge methods implement a common response message pattern:
+
+```proto
+message PurgeBooksResponse {
+  // The number of books that this request deleted (or, if `force` is false,
+  // the number of books that will be deleted).
+  int32 purge_count = 1;
+
+  // A sample of the books that will be deleted.
+  // Only populated if `force` is set to false.
+  repeated string purge_sample = 2;
+}
+```
+
+- The `purge_count` field **should** be included, and provide the number of
+  resources that were deleted (or would be deleted). This count **may** be an
+  estimate (but the service **should** document this if so).
+- The `purge_sample` field **should** be included: If `force` is `false`, it
+  **should** provide a sample of resource names that will be deleted. If
+  `force` is true, this field **should not** be populated.
+
```

Comment:
We should also make it clear that even if both purge_count and purge_sample are missing, force is still required to be explicit in the request that we intend to remove the items matching the filter. Yes, it's a bit repetitive, but we're talking about blasting out potentially thousands of rows in a single HTTP request....